### ISSUE TITLE
CA-based system for client certificates

### ISSUE THREAD

**Issue by slingamn** [oragono/oragono#414]
IMO our current system for handling certfps is not very useful, in that it offers few advantages relative to the use of a strong and unique password over TLS.

My sense of how client certificates are used in the wild is that typically they get issued by some certificate authority, and then a service endpoint (like Oragono) verifies the client certificate against the CA certificate, checks it against a CRL, and then sets the authorized account name to be some field of the certificate (maybe just CN / common name). This has several advantages:

1. Allows multiple certificates per account
1. Allows rotation, expiration, and revocation of certificates per account

The account MUST exist in the database as well, i.e., you can't just present an arbitrary valid certificate and have Oragono create an account for it on the fly if necessary. (These accounts could be created by a one-parameter mode of `SAREGISTER`, i.e., `/msg NickServ SAREGISTER user38475` registers an account that has no passphrase and cannot be logged into via a passphrase, only a valid CA-signed certificate.) Reasons:

1. We want to have a valid account entry for storing user preferences, etc.
1. Revoking access shouldn't require pushing out a new CRL, you should just be able to delete the account and lock them out

TODO: clarify how passphrase-less accounts interact with `NS PASSWD`.

**Comment by DanielOaks** [comment]
This sounds fine. However, it's a really large change in terms of IRC and we will almost certainly be doing something different to all other IRC services software that handles client certs.

While [RFC 7194](https://tools.ietf.org/html/rfc7194#section-2.2.2) does offer some recommendations on client certs, it... feels like these aren't useful. We'll want to write up our own set of recommendations based on how we end up implementing this, which we can push towards v3 for wider adoption.

On the side, how common are CA-authenticated client certificates?

**Comment by DanielOaks** [comment]
Should be noted: This would be an alternative system to the existing defacto standard way using certificate fingerprints.

**Comment by slingamn** [comment]
Actually I think this proposal is fairly close to the RFC 7194 recommendations (which I was not previously aware of, thanks!):

1. "If the client is using a certificate as well, it should be issued by a commonly trusted CA or a CA designated by the IRC network." [This proposal requires a designated CA, but if it's easy to support a bundle of some kind as well, we could do that.]
1. "The certificate's Common Name should match the main IRC nickname." [account name in our case; these concepts are converging for us, particularly with 775ead700ff9948b98f4349cfd4ad2293a23bbfc]

I think this is something that would exist alongside the current mechanism of supporting client certificates, e.g., we'd add a section `cacert` to the `accounts` section of the config, and `cacert: true` would become a new field of the JSON auth blob alongside `passphrasehash` and `certfp`.

I'm interested in learning more about how current service frameworks handle client certificates. I found the [oftc](https://www.oftc.net/NickServ/CertFP/) page on the subject, which sounds similar to Oragono's current behavior.

(re. "really large change", I think this is the same sort of change as the `require-sasl` mode: it's a large change in spirit but only a small one in practice, and one that should be adequately supported by existing client software without changes or much of an impedance mismatch.)

Anecdotally, client certificates on the CA model are popular with enterprises, and with some governments, as a way of implementing single-sign-on. [Vault](https://www.vaultproject.io/docs/auth/cert.html) supports them as a means of authentication to Vault (Vault can also issue them for authentication to other services). I found an article describing their use by the government of [Estonia](https://www.ndss-symposium.org/wp-content/uploads/2017/09/12_4_1.pdf).

Unfortunately, I also found a [source](https://ieeexplore.ieee.org/document/8002897) describing how TLS v1.2 sends the certificate itself in plaintext before the session key has been negotiated, allowing tracking of users by their certificates. So this is an axis along which client certificates are actually worse than tunneling SASL PLAIN over TLS.
